include/zephyr/sys/hash_function.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2022 Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_HASH_FUNCTION_H_
+#define ZEPHYR_INCLUDE_SYS_HASH_FUNCTION_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util_macro.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief 32-bit Hash function interface
+ *
+ * Hash functions are used to map data from an arbitrarily large space to a
+ * smaller, fixed-size space. For a given input, a hash function will
+ * consistently generate the same, semi-unique numerical value. Even for
+ * marginally different data, a good hash function will distribute the entropy
+ * almost evenly over all bits in the hashed value. When combined with modulu
+ * arithmetic over a finite-sized numeric field, a good hash function will
+ * produce a nearly uniform distribution.
+ *
+ * @param str a string of input data
+ * @param n the number of bytes in @p str
+ *
+ * @return the numeric hash associated with @p str
+ */
+typedef uint32_t (*sys_hash_func32_t)(const void *str, size_t n);
+
+/**
+ * @brief Daniel J. Bernstein's hash function
+ *
+ * Some notes:
+ * - normally, this hash function is used on NUL-terminated strings
+ * - it has been modified to support arbitrary sequences of bytes
+ * - it has been modified to use XOR rather than addition
+ *
+ * @param str a string of input data
+ * @param n the number of bytes in @p str
+ *
+ * @return the numeric hash associated with @p str
+ *
+ * @note enable with @kconfig{CONFIG_SYS_HASH_FUNC32_DJB2}
+ *
+ * @see https://theartincode.stanis.me/008-djb2/
+ */
+uint32_t sys_hash32_djb2(const void *str, size_t n);
+
+/**
+ * @brief System default 32-bit hash function
+ *
+ * @param str a string of input data
+ * @param n the number of bytes in @p str
+ *
+ * @return the numeric hash associated with @p str
+ */
+static inline uint32_t sys_hash32(const void *str, size_t n)
+{
+	if (IS_ENABLED(CONFIG_SYS_HASH_FUNC32_CHOICE_DJB2)) {
+		return sys_hash32_djb2(str, n);
+	}
+
+	__ASSERT(0, "No default 32-bit hash. See CONFIG_SYS_HASH_FUNC32_CHOICE");
+
+	return 0;
+}
```

Comment:
how does this work if the user wants to add its own custom hashing function? Can you extend the test to cover this case?

Reply:
This inline is mainly for the default, system-wide 32-bit hash function.

If a user wants to use a custom hash function for a Hashmap, then it needs to be a part of the `sys_hashmap.api`. See `SYS_HASHMAP_DEFINE_ADVANCED()`. I could probably document that better.